[Final Submission] Battlezone

end-ichi opened the pull request:
PR for the integration of Battlezone into the main JAXAtari repo. More changes to follow.

IIITyforIII committed:
try out lost luggage

IIITyforIII committed:
swap to battlezone

end-ichi committed:
changed filename of experimental fork

xxxale committed:
Add version v5 to 'ALE/BattleZone'

IIITyforIII committed:
check individual pixels by plotting observation

IIITyforIII committed:
added basic structure

IIITyforIII committed:
step_counter implemented and black layout added

IIITyforIII committed:
added sprite extractor tool

end-ichi committed:
ram viewer

IIITyforIII committed:
added sprites and ways to display them

IIITyforIII committed:
added anim counters for chain

IIITyforIII committed:
chain animation works now

IIITyforIII committed:
mountain scrolling works now + numerical optimization

IIITyforIII committed:
grass front animation works now

xxxale committed:
add basic enemy logic

IIITyforIII committed:
merge conflict avoidance

IIITyforIII committed:
fixed self.consts and added initial radar spin

IIITyforIII committed:
added radar circle

IIITyforIII committed:
added life

IIITyforIII committed:
pushing incomplete score code to avoid merge conflicts

IIITyforIII committed:
not beatuiful but score works i guess (cant make number of digits dynamic without major changes and not important)

end-ichi committed:
Improved RAM viewer and added log parser

- figured out some of the RAM addresses' purpose
- added ability to tickstep when paused
- added ability to log certain addresses every frame
- added script to analyze those logs

end-ichi committed:
new game details + RAM viewer improvements

xxxale committed:
Rearrange _enemy_step logic

xxxale committed:
fix enemy offset_xy

xxxale committed:
Enemy position change w respect to player

xxxale committed:
Add point for enemies in radar

xxxale committed:
fix enemy point only displayed if in radar radius

xxxale committed:
fix movement right or left to rotation not translation

xxxale committed:
use RADAR_MAX_SCAN_RADIUS as constant not argument

xxxale committed:
add PLAYER_SPEED + rename PLAYER_ROTATION_SPEED

xxxale committed:
remove print statement

IIITyforIII committed:
extracted some enemy images

IIITyforIII committed:
created skeleton for enemy render

IIITyforIII committed:
added first zoom feature but its completely scuffed for now

IIITyforIII committed:
"fixed" zoom still does not look exactly like the original but i think that would go to the point of replicating artifacts etc. which would be overkill imo

xxxale committed:
remove ToDo + fix player (rotation) speed

IIITyforIII committed:
fixed zoom to work with dynamic distances

IIITyforIII committed:
zoom now works with dynamic distance

IIITyforIII committed:
Merge branch 'master' into HEAD

# Conflicts:
#	src/jaxatari/games/jax_battlezone.py

IIITyforIII committed:
adjusted some values

IIITyforIII committed:
added provisorical screen cords mapping

IIITyforIII committed:
found my mistake screen cord mapping works now and makes sense

IIITyforIII committed:
magic number removal

IIITyforIII committed:
reverted environment.py changes... we are not supposed to change that file i think

IIITyforIII committed:
reworked zoom to work with padded enemy masks

IIITyforIII committed:
removal of todos

IIITyforIII committed:
added rotation sprite select code.
(currently works by having both the flipped and normal variant in memory / pad after flipping ofc this is memory inefficient but flipping after pad is painful)

xxxale committed:
New player speed

xxxale committed:
fix dual movements of player like UpRight

end-ichi committed:
reverse engineering stuff

IIITyforIII committed:
some sprite extraction

xxxale committed:
add avtive filed for enemy

xxxale committed:
fix only position of active enemies is updated

xxxale committed:
add second enemy

xxxale committed:
fix only show active enemies in radar

xxxale committed:
fix only render active enemies

IIITyforIII committed:
Revert "some sprite extraction"

This reverts commit b7e1972915287f5024effb5e04915b138d1048d6.

IIITyforIII committed:
some sprite extraction(was on wrong branch before my bad)

IIITyforIII committed:
projectiles first draft

IIITyforIII committed:
i think rotation is fixed

IIITyforIII committed:
rotation was not fixed and this is the best i can do ;(

IIITyforIII committed:
Hitboxes first draft

IIITyforIII committed:
hitboxes work now

IIITyforIII committed:
bugfix projectiles

IIITyforIII committed:
added enemy explosions

IIITyforIII committed:
better bullet render + improved hitbox

xxxale committed:
add spwan enemies

xxxale committed:
add spawn probabilities

IIITyforIII committed:
fixed death anim that was broken because of spawning

IIITyforIII committed:
fixed saucer sprite

xxxale committed:
add done when player out of lifes

xxxale committed:
add method for when player got shot

xxxale committed:
fix performance of _position_update

xxxale committed:
reduce area of spawning

IIITyforIII committed:
fixed rotated sprite selection

xxxale committed:
add basic movement for enemies

IIITyforIII committed:
fixed enemy movement

IIITyforIII committed:
enemy projectiles now work

xxxale committed:
fix enemy spawn after shot with 0 points

xxxale committed:
add phase to enemies

xxxale committed:
add saucer movement

xxxale committed:
fix rot_centre with correct point

xxxale committed:
fix enemy speed change

end-ichi committed:
added targeting indicator

end-ichi committed:
changed some constants

- halved radar speed so it is roughly the same as the original when running at 60fps
- enemy speed and turn speed

end-ichi committed:
changed more constants and added some comments

end-ichi committed:
fix player fire cooldown not reset on death

end-ichi committed:
fixed death animation was fixed to screen position, not world position

end-ichi committed:
made world toroidal

end-ichi committed:
added projectile despawning

end-ichi committed:
organized constants

end-ichi committed:
mostly fixed action parsing

UPLEFTFIRE and DOWNRIGHTFIRE are not registered for some reason but we can't do anything about that.

end-ichi committed:
changed tank min distance to player

as observed in original

xxxale committed:
fix tank movement

end-ichi committed:
comments

xxxale committed:
fix saucer movement

xxxale committed:
Merge remote-tracking branch 'origin/battlezone' into battlezone

# Conflicts:
#	src/jaxatari/games/jax_battlezone.py

xxxale committed:
fix tank movement

end-ichi committed:
remove output_rgba.npy

end-ichi committed:
hopefully fixing observation space shape error

end-ichi committed:
fixing the tests

end-ichi committed:
added correct turn+drive speed

IIITyforIII committed:
adjusted some values

shaikgermany2105-coder commented:
<img width="516" height="861" alt="Screenshot 2026-01-04 113357" src="https://github.com/user-attachments/assets/2263c576-3453-4757-96c1-b3cb215fb6d4" />




<img width="448" height="695" alt="Screenshot 2026-01-04 100902" src="https://github.com/user-attachments/assets/ef928555-4e90-49d1-a044-881d30ea6671" />



hallo,I noticed that there are two implementations of Battlezone, and I wanted to clarify which one you would like me to review. Please let me know.

IIITyforIII commented:
> hallo,I noticed that there are two implementations of Battlezone, and I wanted to clarify which one you would like me to review. Please let me know.

Hi, our version is in the file jax_battlezone.(the second picture i think)
(jax_battlezone_experimantelfork is from the previous team that worked on battlezone.  battlezone_dev is just for testing/reverse engineering)

IIITyforIII committed:
last commit i broke enemy projectiles ;( (see review) this is fixed now / only worked before because enemy hitboxes where wrong)

shaikgermany2105-coder commented:
Review for Battlezone

ALE-Similarity: 3/4
The environment captures many core mechanics of ALE Battlezone reasonably well, including toroidal world wrapping, player movement/rotation (with speed adjustments for diagonal inputs), projectile firing/cooldowns, and basic enemy AI (e.g., Tank maintaining distance and turning toward the player). Radar rendering and enemy spawning logic (with score-based probabilities) also align closely.
However, what I observed is that enemy death animations and some timings (e.g., radar rotation speed) may not perfectly sync.
Some enemy behaviors and the zoom/perspective differ from the ALE version.
There are also quite a few TODOs, which gives the impression that some areas are still in progress.

Implementation quality: 4/4
JAX is effectively utilized for performance, with jax.jit, vmap, and lax.cond enabling efficient vectorized enemy/projectile updates and rendering scans, avoiding sequential loops and no noticeable lags.Overall it already feels close to the original game. With some refinement in enemy behavior and code structure, it could become a very strong ALE-style reimplementation.

end-ichi committed:
fixed some saucer movement bugs

- stored points for saucer movement are now updated same as enemy position when player moves
- added world wrapping in move_to_player function
- added world wrapping in saucer movement function
- added world wrapping for stored points in saucer movement when saucer position wraps (wrapping = saucer is "moved" by WORLD_SIZE along X or Z axis)

end-ichi committed:
changed hitbox and render position + saucer improvement

- smaller hitbox to be more like original
- fixed enemy sprites being rendered at incorrect position (now centered on position rather than to the side)
- same for death animation
- also adapted enemy/projectile collision logic for centered positioning/rendering
- changed player hitbox to reflect original (shifted into positive z)
- improved saucer movement (but still not great) by strafing to avoid player aim.
- also adapted target indicator

end-ichi committed:
new sprite centering logic now also for projectiles

xxxale committed:
expand observation space + adjust observation parts

xxxale committed:
remove obs_size from init

end-ichi committed:
more log_parse stuff

end-ichi committed:
added fighter jet

also fixed some saucer bugs

IIITyforIII committed:
fixed figher jet death anim

IIITyforIII committed:
Merge branch 'battlezone'

IIITyforIII committed:
Merge remote-tracking branch 'upstream/master'

IIITyforIII committed:
changed to new transparency color index

IIITyforIII committed:
found transparency_id already exists in rendering utils so i replaced it

IIITyforIII committed:
fixed all remaining bugs introduced by the upstream master merge

IIITyforIII committed:
implemented new conventions for constants and states:
States, observations, and info must now be implemented as flax.struct.dataclasses.
Constants should be defined as flax.struct.PyTreeNode.

IIITyforIII committed:
implemented new conventions for object observations

IIITyforIII committed:
removed imports that my ide put in for some reason

IIITyforIII committed:
removed some todos

IIITyforIII committed:
restructured imports

IIITyforIII committed:
fixed observation space

IIITyforIII committed:
added some mods

IIITyforIII committed:
fixed mods

xxxale committed:
adjustment from 60fps to 30fps

end-ichi committed:
Bugfixes (Saucer movement, world wrap)

- proper world wrapping for enemy position as well as stored points (these wrap when position wraps)
- fixed saucer strafing still using old hitbox positioning
- some refactoring

- changed saucer phase after being near_player. going straight into rotation phase caused a bug where the saucer would encircle the player forever. my best guess is that it was caused by the saucer exiting near_player, then set its new rotation center in p1, rotation around which immediately brought it back into near_player (and this looping forever). the rotation center was constatntly shifting even with player stationary and this is the only explanation I could come up with for that. it seems fixed now, at the cost of the saucer being a little easier to kill.

end-ichi committed:
code style enhancements

- improved readability
- removed comment used for gathering information in the beginning
- removed commented-out debug prints

end-ichi committed:
fixing pipeline tests

end-ichi committed:
chenged constants + more pipieline test fixes

- changed bg scroll speed and FOV to be more like original
- deleted some files and added one to make the pipeline tests stop complaining about them

end-ichi committed:
fixing object space pipeline test failures

negative values are apparently disallowed in object space

end-ichi committed:
fixing more observation bugs

- removed some stuff I missed deleting during the previous observation bugfix
- added clipping to the enemy and projectile y to adhere to object observation schema (doesn't allow off-screen values)
- ensured player projectile state conversion to 1d array for concatenation

end-ichi committed:
fixing even more observation bugs

- added clipping to all x and y values for object observations

end-ichi committed:
fixed enemy rendering order

- enemies are now rendered in order of distance (furthest first)

IIITyforIII committed:
pushed best working reward func for now
(no delta in reward)

xxxale committed:
add supertank with basic tank movement

xxxale committed:
add more realistic supertank movement

xxxale committed:
remove completed ToDo comments in supertank

xxxale committed:
change saucer min_dist to const

end-ichi committed:
Saucer despawn

- when the saucer is too far away it despawns now
- this is to fix the case where there are two saucers far away and the player having no idea where they are and just driving through a seemingly empty map
- in the original, this also appears to be implemented, though it's difficult to pin down specifics without a lot of effort.

end-ichi committed:
fix merge conflict bug

end-ichi committed:
adjust saucer strafe speed

- became too jerky since FOV was narrowed

end-ichi committed:
removing superfluous files

end-ichi committed:
revert .gitignore

end-ichi committed:
added newlines to end of gitignore

- original gitignore ends in two newlines

IIITyforIII committed:
added enemy friendly fire

IIITyforIII committed:
deleted superfluos code
